pkg/collector/pipeline.go

```diff
@@ -49,17 +49,41 @@ func (c nowClock) Now() time.Time {
 var defaultClock nowClock
 
 // Pipeline is a series of processors that should be applied to each report that
-// the collector receives.
+// the collector receives. It uses a fixed number of workers to process the reports
+// and a fixed sized queue that the workers read from. If the queue fills, reports
+// are dropped.
 type Pipeline struct {
 	processors []ReportProcessor
 	clock      Clock
+	c          chan *ReportBatch
 }
 
-// NewPipeline creates a new Pipeline that uses a particular clock.  For
-// production pipelines, just instantiate the Pipeline type yourself
-// (&Pipeline{}).
-func NewPipeline(clock Clock) *Pipeline {
-	return &Pipeline{clock: clock}
+// NewPipeline creates a new Pipeline with a specified buffer size.
+func NewPipeline(bufferSize int64, numWorkers int) *Pipeline {
+	return setupPipeline(context.Background(), nil, bufferSize, numWorkers)
+}
+
+const defaultBufferSize = 1000
+const defaultNumWorkers = 10
+
+// NewTestPipeline creates a new Pipeline with a specified clock.
+// This should only be used for testing.
+func NewTestPipeline(clock Clock) *Pipeline {
+	return setupPipeline(context.Background(), clock, defaultBufferSize, defaultNumWorkers)
+}
+
+// NewTestPipelineWithBuffer creates a new Pipeline with a specified buffer size and clock.
+// This should only be used for testing.
+func NewTestPipelineWithBuffer(clock Clock, bufferSize int64) *Pipeline {
+	return setupPipeline(context.Background(), clock, bufferSize, defaultNumWorkers)
+}
+
+func setupPipeline(ctx context.Context, clock Clock, bufferSize int64, numWorkers int) *Pipeline {
+	p := &Pipeline{clock: clock, c: make(chan *ReportBatch, bufferSize)}
+	for i := 0; i < numWorkers; i++ {
+		go p.runPipeline(ctx)
```

Comment:
Are you planning to add some way of cancelling / cleaning these up in a follon PR? Seems important.

Reply:
On the one hand, this is for something that should just run forever, so I'm not clear that it's super important, but I have added a Close method that does so, which will be helpful for tests